### PR TITLE
[Fix #1337] Update the cop to raise error for insufficient descriptions.

### LINKED ISSUE #1337: Cop Idea: prevent "it works" description
I see too many this:

```ruby
it "works" do
  # ...
end
```

Specs are not just here to prevent bugs, but are, as the name implies, specifications. In other words, a documentation we can use as developer to understand what the code is supposed to do.

So it could be nice to enforce a description of "what it does". Preventing "it works" would be a good start 😋 

```ruby
# bad

it "works" do
  # ...
end

# good

it "marks the task as done" do
  # ...
end
```

### PR DESCRIPTION
**Pull request by akrox58**
Per issue #1337, I tried to add a generic way to the ExampleWordingCop to call out insufficient descriptions. For now, I've used "works" and "throws errors" as that is a phrase I see used often as well. I didn't use a new cop since insufficient descriptions seemed to be along the same lines as "should not start with should or it" as they are both description related.

---

Before submitting the PR make sure the following are checked:

* [X] Feature branch is up-to-date with `master` (if not - rebase it).
* [X] Squashed related commits together.
* [X] Added tests.
* [X] Updated documentation.
* [X] Added an entry to the `CHANGELOG.md` if the new code introduces user-observable changes.
* [X] The build (`bundle exec rake`) passes (be sure to run this locally, since it may produce updated documentation that you will need to commit).

If you have modified an existing cop's configuration options:

* [X] Set `VersionChanged` in `config/default.yml` to the next major version.

### CODE REVIEW

File: lib/rubocop/cop/rspec/example_wording.rb
Comment (ydah) [inline lib/rubocop/cop/rspec/example_wording.rb:60]: It would be better if it could be changed with a configurable option, as suggested in https://github.com/rubocop/rubocop-rspec/issues/1337#issuecomment-1196650696

File: lib/rubocop/cop/rspec/example_wording.rb
Comment (bquorning) [inline lib/rubocop/cop/rspec/example_wording.rb:76]:
```suggestion
            elsif INSUFFICIENT_EXAMPLES.include?(message.strip.squeeze(' '))
```

File: spec/rubocop/cop/rspec/example_wording_spec.rb
Comment (pirj) [inline spec/rubocop/cop/rspec/example_wording_spec.rb:195]: Those should be examples of what the cop should ignore.
The full message is "throws error with a message". Why is this an offence?

It seems that:
 - we don't consider articles e.g. "throws an error"
 - we don't inspect complete messages, but just the first part of it

**Review by pirj — CHANGES_REQUESTED** [review]

File: lib/rubocop/cop/rspec/example_wording.rb
Comment (akrox58) [reply lib/rubocop/cop/rspec/example_wording.rb:76]: Thank you for the correction!

File: lib/rubocop/cop/rspec/example_wording.rb
Comment (akrox58) [reply lib/rubocop/cop/rspec/example_wording.rb:60]: DisallowedExamples:
    - works
    
My concern with adding it as a configurable option is the possibility of how big the list can get 😅 Are we anticipating it to get really big eventually?

File: spec/rubocop/cop/rspec/example_wording_spec.rb
Comment (akrox58) [reply spec/rubocop/cop/rspec/example_wording_spec.rb:195]: Thanks for bringing up the article part! you are right that we don't consider articles! I went down a rabbit hole of trying to ignore the articles and strip the phrase to just ["throws error"] from ["throws error, throws an error"] etc... But then I realized there could be grammatical errors that get missed from it. so I'm a little confused at what to do. I removed the phrase for the time being though. 

as for #2, I thought include? checks for whole phrases. so we will ignore "throws an error with a message" but the cop will sound an alarm for "throws error". Is there something I'm missing?

File: lib/rubocop/cop/rspec/example_wording.rb
Comment (ydah) [reply lib/rubocop/cop/rspec/example_wording.rb]: I think each project would have different examples of settings you want to disallowed, so wouldn't it be better to allow each project to configure its own settings?

File: lib/rubocop/cop/rspec/example_wording.rb
Comment (ydah) [inline lib/rubocop/cop/rspec/example_wording.rb:34]: It would be good to have a good case/bad case according to the setting values of the configuration options.

File: lib/rubocop/cop/rspec/example_wording.rb
Comment (ydah) [inline lib/rubocop/cop/rspec/example_wording.rb:135]: If I format it with `downcase`, won't it always mismatch if there are uppercase letters in the `DisallowedExample` setting?

Comment (akrox58) [reply lib/rubocop/cop/rspec/example_wording.rb:135]: good callout! we should not allow uppercase in the DisallowedExample setting! Updating!

Comment (akrox58) [reply lib/rubocop/cop/rspec/example_wording.rb:140]: great callout. I've added a comment that the words in this config will not be treated as case sensitive!

File: lib/rubocop/cop/rspec/example_wording.rb
Comment (akrox58) [reply lib/rubocop/cop/rspec/example_wording.rb:38]: I think I addressed it with my latest changes, but let me know if you were expecting something else!

File: lib/rubocop/cop/rspec/example_wording.rb
Comment (akrox58) [reply lib/rubocop/cop/rspec/example_wording.rb]: ah gotcha!!! makes sense!!

File: lib/rubocop/cop/rspec/example_wording.rb
Comment (ydah) [reply lib/rubocop/cop/rspec/example_wording.rb:140]: I think it looks good.
Also, would it look good to add the following test?

```ruby
context 'when `DisallowedExamples: Workz`' do
  let(:cop_config) { { 'DisallowedExamples' => ['Workz'] } }

  it 'flags an unclear description' do
    expect_offense(<<-'RUBY')
      it "workz" \
          ^^^^^^^^ Your test should be more descriptive.
        "with #{object}" do
      end
    RUBY
  end

  it 'flags an unclear description despite extra spaces' do
    expect_offense(<<-'RUBY')
      it "  workz    " \
          ^^^^^^^^^^^^^^ Your test should be more descriptive.
        "with #{object}" do
      end
    RUBY
  end

  it 'flags an unclear description despite uppercase and lowercase strings' do
    expect_offense(<<-'RUBY')
      it "WOrKz " \
          ^^^^^^^^^ Your test should be more descriptive.
        "with #{object}" do
      end
    RUBY
  end
end
```

File: lib/rubocop/cop/rspec/example_wording.rb
Comment (ydah) [reply lib/rubocop/cop/rspec/example_wording.rb:38]: I think the part you added is an explanation of the case where 'it' is repeated. I thinking of the https://github.com/rubocop/rubocop-rspec/pull/1350#discussion_r938271855 changes. For your information.

File: lib/rubocop/cop/rspec/example_wording.rb
Comment (akrox58) [reply lib/rubocop/cop/rspec/example_wording.rb:45]: Ahhh sorry typo :/

File: lib/rubocop/cop/rspec/example_wording.rb
Comment (akrox58) [reply lib/rubocop/cop/rspec/example_wording.rb:140]: @ydah  I thought I saw a notification on updating the test cases! that was a good idea! have added that too! 
Appreciate your patience on this! 
PR is ready for review again 😅

Comment (ydah) [reply lib/rubocop/cop/rspec/example_wording.rb:140]: Oops, I thought I had added a comment about the proposed test, but it disappeared. Thank you so much!

**Review by ydah — APPROVED** [review]
Looks good to me, thank you.

**Comment by akrox58** [comment]
Thanks @ydah! How do I get this merged? Do I need to wait for the other reviewers to approve?

**Review by pirj — CHANGES_REQUESTED** [review]
I would expect the following to pass, but it fails, why?
```ruby
    expect_no_offenses(<<-'RUBY')
      it "works" \
        "totally fine" do
      end
    RUBY
```

**Comment by akrox58** [comment]
> I would expect the following to pass, but it fails, why?
> 
> ```ruby
>     expect_no_offenses(<<-'RUBY')
>       it "works" \
>         "totally fine" do
>       end
>     RUBY
> ```

Looks like I didn't handle that edge case! updated the check to check for the entire description instead, and added some more test cases! Thanks for the catch!

**Review by pirj — APPROVED** [review]
I've checked against https://github.com/pirj/real-world-rspec, no errors, and the cop "just works" 😄 

```
...
/Users/pirj/source/real-world-rspec/sharetribe/spec/models/person_spec.rb:338:9: C: [Correctable] RSpec/ExampleWording: Your test should be more descriptive.
    it 'works' do
        ^^^^^
/Users/pirj/source/real-world-rspec/solidus/core/spec/lib/spree/core/testing_support/factories/address_factory_spec.rb:41:11: C: [Correctable] RSpec/ExampleWording: Your test should be more descriptive.
      it 'works' do
          ^^^^^

57497 files inspected, 273 offenses detected, 273 offenses autocorrectable
```

Looks great, thank you for the contribution!

File: config/default.yml
Comment (akrox58) [reply config/default.yml:372]: @pirj Is there something I would need to do to address this?

Comment (pirj) [reply config/default.yml:374]: No, it is fine. Just thinking out loud to get an opinion of maintainers.
For me, this change is totally fine as is.

**Comment by bquorning** [comment]
I have renamed some of the required CI checks, which unfortunately means that you will need to rebase this branch to make CI pass.

**Comment by pirj** [comment]
Rebased and green @bquorning @Darhazer

File: CHANGELOG.md
Comment (bquorning) [inline CHANGELOG.md:41]: Could you move this up to the `(Unreleased)` section of the Changelog, please?

File: lib/rubocop/cop/rspec/example_wording.rb
Comment (bquorning) [inline lib/rubocop/cop/rspec/example_wording.rb:141]: Won't this line allocate (at least) 3 String objects on each `on_block` call? I don’t know if it would be possible to memoize this at class level.

**Review by bquorning — COMMENTED** [review]
> This cop will also look for insufficient examples and call them out.

> Use the DisallowedExamples setting to prevent unclear or insufficient descriptions.

> `MSG_INSUFFICIENT_DESCRIPTION = 'Your test should be more descriptive.'`

I think we use a more clear terminology. Internally in rspec-core, the first argument to `it` is called `desc`. So perhaps

> This cop will also look for insufficient example descriptions and call them out.

> Use the DisallowedExampleDescriptions setting to prevent unclear or insufficient example descriptions.

> `MSG_INSUFFICIENT_DESCRIPTION = 'Your example description is insufficient.'`

File: lib/rubocop/cop/rspec/example_wording.rb
Comment (pirj) [reply lib/rubocop/cop/rspec/example_wording.rb]: Since, in theory, different `DisallowedExamples` can be configured in different directories, it would be incorrect to cache this at the class level.
I've made the change to cache it in an instance variable, as I guess the same instance will be used for a single investigation.
We also need to [nullify the instance variable in `on_new_investigation`](https://docs.rubocop.org/rubocop-rspec/upgrade_to_version_2.html#always-call-super-from-on_new_investigation-in-your-cops). Is it worth it?

Comment (bquorning) [reply lib/rubocop/cop/rspec/example_wording.rb]: No, probably not worth it. Thanks for investigating.

**Review by bquorning — APPROVED** [review]

**Comment by pirj** [comment]
Thank you for the contribution and your time, @akrox58 !

**Pull request merged**
